src/Language/PureScript/TypeChecker/Unify.hs

```diff
@@ -5,53 +5,77 @@
 --
 module Language.PureScript.TypeChecker.Unify
   ( freshType
+  , freshTypeWithKind
   , solveType
   , substituteType
   , unknownsInType
   , unifyTypes
   , unifyRows
   , alignRowsWith
-  , replaceVarWithUnknown
   , replaceTypeWildcards
   , varIfUnknown
   ) where
 
 import Prelude.Compat
 
-import Control.Arrow (first, second)
 import Control.Monad
 import Control.Monad.Error.Class (MonadError(..))
-import Control.Monad.State.Class (MonadState(..), gets, modify)
+import Control.Monad.State.Class (MonadState(..), gets, modify, state)
 import Control.Monad.Writer.Class (MonadWriter(..))
 
 import Data.Foldable (traverse_)
-import Data.Function (on)
-import Data.List (sortBy, nubBy)
 import qualified Data.Map as M
-import Data.Ord (comparing)
-import Data.Text (Text)
 import qualified Data.Text as T
 
 import Language.PureScript.Crash
+import qualified Language.PureScript.Environment as E
 import Language.PureScript.Errors
+import Language.PureScript.TypeChecker.Kinds (elaborateKind, instantiateKind, unifyKinds)
 import Language.PureScript.TypeChecker.Monad
 import Language.PureScript.TypeChecker.Skolems
 import Language.PureScript.Types
 
--- | Generate a fresh type variable
+-- | Generate a fresh type variable with an unknown kind. Avoid this if at all possible.
 freshType :: (MonadState CheckState m) => m SourceType
-freshType = do
-  t <- gets checkNextType
-  modify $ \st -> st { checkNextType = t + 1 }
-  return $ srcTUnknown t
+freshType = state $ \st -> do
+  let
+    t = checkNextType st
+    st' = st { checkNextType = t + 2
+             , checkSubstitution =
+                 (checkSubstitution st) { substUnsolved = M.insert t (UnkLevel (pure t), E.kindType)
+                                                        . M.insert (t + 1) (UnkLevel (pure (t + 1)), srcTUnknown t)
+                                                        . substUnsolved
+                                                        $ checkSubstitution st
+                                        }
+             }
+  (srcTUnknown (t + 1), st')
+
+-- | Generate a fresh type variable with a known kind.
+freshTypeWithKind :: (MonadState CheckState m) => SourceType -> m SourceType
+freshTypeWithKind kind = state $ \st -> do
+  let
+    t = checkNextType st
+    st' = st { checkNextType = t + 1
+             , checkSubstitution =
+                 (checkSubstitution st) { substUnsolved = M.insert t (UnkLevel (pure t), kind) (substUnsolved (checkSubstitution st)) }
+             }
+  (srcTUnknown t, st')
 
 -- | Update the substitution to solve a type constraint
 solveType :: (MonadError MultipleErrors m, MonadState CheckState m) => Int -> SourceType -> m ()
 solveType u t = do
   occursCheck u t
+  k1 <- elaborateKind t
+  subst <- gets checkSubstitution
+  let k2 = substituteType subst
+         . maybe (internalError ("No kind for unification variable ?" <> show u)) snd
```

Comment:
I don't think it's always possible to use a monadic internal compiler error, because we often use `internalError` in "pure" functions. This one could be, but I also wonder if we should just make this a "real" exception which we catch via IO and turn into a compiler error.

Reply:
If that's even possible.

Reply:
Can you expand on that? I’m sure I’m missing something here but to me this seems like a good candidate for `internalCompilerError`, since we already have the MonadError MultipleErrors constraint?

Reply:
Sorry, that was more in response to

> and to create a new issue for changing the existing uses of internalError to use internalCompilerError where we are inside a monad that supports it?

I'm wondering if it's actually better to make a new `InternalError` exception data type that we catch and rethrow in places that would preserve the context. I don't know if we can turn all usages of `internalError` into `internalCompilerError` in general, and so maybe it can work such that we don't need to. I don't know though since laziness is always an issue.

Reply:
My initial reaction is that the fewer different channels we have for different kinds of errors, the better (probably), and I don't understand what the problem with using the normal MonadError MultipleErrors channel is. What properties of MonadError make it less suitable, or what makes exceptions more suitable?

Reply:
Just that I think there are places we use `internalError` that aren't `MonadError MultipleErrors`. I could be wrong though.

Reply:
Yeah there are a few, there's one in `moduleToExternsFile`, and there's another in `renderSimpleErrorMessage` (although we could probably get rid of that one by using NonEmptyList in the OverlappingInstances constructor). Anyway, I suspect we are in `MonadError MultipleErrors` in the majority of cases where `internalError` is used, and by using the `MonadError MultipleErrors` channel, the error will always contain the PureScript module in which the error was thrown without us having to do any extra work, right? I think having a separate internal error function for the few cases where we aren't in `MonadError MultipleErrors` is an acceptable price to pay for that.